File: src/test/java/javasolutions/arraysandhashing/TopKFrequentElementsTest.java
package javasolutions.arraysandhashing;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class TopKFrequentElementsTest {

    @Test
    void should_return_array_with_two_elements() {
        // given
        int[] nums = new int[]{1, 1, 1, 2, 2, 3};
        int k = 2;
        TopKFrequentElements topKFrequentElements = new TopKFrequentElements();
        // when
        int[] result = topKFrequentElements.topKFrequent(nums, k);
        // then
        assertArrayEquals(result, new int[]{1, 2});
    }

    @Test
    void should_return_array_with_one_element() {
        // given
        int[] nums = new int[]{1};
        int k = 1;
        TopKFrequentElements topKFrequentElements = new TopKFrequentElements();
        // when
        int[] result = topKFrequentElements.topKFrequent(nums, k);
        // then
        assertArrayEquals(result, new int[]{1});
    }

}

File: src/test/kotlin/kotlinsolutions/arraysandhashing/TopKFrequentElementsTest.kt
package kotlinsolutions.arraysandhashing

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class TopKFrequentElementsTest {

    @Test
    fun should_return_array_with_two_elements() {
        // given
        val nums = intArrayOf(1, 1, 1, 2, 2, 3)
        val k = 2;
        val topKFrequentElements = TopKFrequentElements()
        // when
        val firstResult = topKFrequentElements.topKFrequent(nums, k)
        val secondResult = topKFrequentElements.topKFrequentOptimal(nums, k)
        // then
        assertArrayEquals(firstResult, intArrayOf(2, 1))
        assertArrayEquals(secondResult, intArrayOf(1, 2))
    }

    @Test
    fun should_return_array_with_one_element() {
        // given
        val nums = intArrayOf(1)
        val k = 1
        val topKFrequentElements = TopKFrequentElements()
        // when
        val firstResult = topKFrequentElements.topKFrequent(nums, k)
        val secondResult = topKFrequentElements.topKFrequentOptimal(nums, k)
        // then
        assertArrayEquals(firstResult, intArrayOf(1))
        assertArrayEquals(secondResult, intArrayOf(1))
    }

    @Test
    fun should_return_array_with_one_negative_value() {
        // given
        val nums = intArrayOf(-1, -1)
        val k = 1
        val topKFrequentElements = TopKFrequentElements()
        // when
        val firstResult = topKFrequentElements.topKFrequent(nums, k)
        val secondResult = topKFrequentElements.topKFrequentOptimal(nums, k)
        // then
        assertArrayEquals(firstResult, intArrayOf(-1))
        assertArrayEquals(secondResult, intArrayOf(-1))
    }

    @Test
    fun should_return_array_with_two_values() {
        // given
        val nums = intArrayOf(4,1,-1,2,-1,2,3)
        val k = 2
        val topKFrequentElements = TopKFrequentElements()
        // when
        val firstResult = topKFrequentElements.topKFrequent(nums, k)
        val secondResult = topKFrequentElements.topKFrequentOptimal(nums, k)
        // then
        assertArrayEquals(firstResult, intArrayOf(-1, 2))
        assertArrayEquals(secondResult, intArrayOf(-1, 2))
    }
}